angular/src/lib/pagination/pagination.component.ts

```diff
@@ -0,0 +1,177 @@
+    <md-pagination-number *ngFor='let item of renderGroupMid'
+      [index]='item'
+      [ifPreventDefault]='ifPreventDefault'
+    ></md-pagination-number>
+    <li class='pagination_li ellipsis' *ngIf='lastGroupStart<this.total' aria-hidden="true">...</li>
+    <md-pagination-number *ngFor='let item of renderGroupLast'
+      [index]='item'
+      [ifPreventDefault]='ifPreventDefault'
+    ></md-pagination-number>
+    <md-pagination-arrow
+      [isPrevious]='false'
+      [ifPreventDefault]='ifPreventDefault'
+    ></md-pagination-arrow>
+  `,
+  styleUrls: ['pagination.scss'],
+  host: {
+    class: 'pagination',
+    'style': `
+      display: block;
+      min-height: 1.5rem;
+      list-style-position: outside;
+      clear:both;
+    `,
+    'attr.role': 'navigation',
+    'attr.aria-label': 'Pagination',
+    'tabindex': '0',
+    '(keydown)': 'handleKeydown($event)'
+  },
+  providers: [PaginationService]
+})
+export class PaginationComponent implements OnInit {
+
+  /** @option set the href format | '' */
+  @Input() href: string = '';
+  /** @option set the replace placeholder of href */
+  @Input() hrefReplaceReg: string | RegExp = '$page$';
+  /** @props set total | '' */
+  @Input() total: number = 10;
+  /** @props set current | '' */
+  @Input() current: number = 1;
+  /** @props set the first button group | '' */
+  @Input() firstGroupNum: number = 2;
+  /** @props set the mid button group | '' */
+  @Input() midGroupNum: number = 3;
+  /** @props set the last button group | '' */
+  @Input() lastGroupNum: number = 2;
+  /** @option Callback function invoked when user clicks buttons */
+  @Output() whenClick = new EventEmitter();
+
+  public renderGroupFirst = [];
+  public renderGroupMid = [];
+  public renderGroupLast = [];
+  public renderGroup = [];
+
+  public firstGrounpEnd: number = -1;
+  public midGroupStart: number = 1;
+  public midGroupEnd: number = 1;
+  public lastGroupStart: number = Infinity;
+
+  public ifPreventDefault = false;
+
+  private minCurrentForFirstGroup = 0;
+  private maxCurrentForLastGroup = 0;
+
+  constructor(
+    private paginationService: PaginationService
+  ) { }
+
+  ngOnInit() {
+    const s = this.paginationService;
+    s.initConfig(this.href, this.hrefReplaceReg, this.total);
+    s.current$.subscribe((v) => {
+      this.build();
+      this.whenClick.emit(v);
+    });
+    this.paginationService.select(this.current);
+  }
+
+  public handleKeydown = (event) => {
+    const s = this.paginationService;
+    let ifStop = false;
+
+    switch (event.keyCode) {
+      case 37: // <-
+        s.focusInList(this.renderGroup, -1);
+        ifStop = true;
+        break;
+      case 39: // ->
+        s.focusInList(this.renderGroup, 1);
+        ifStop = true;
+        break;
+      case 13: // Enter
+        s.selectfocus();
+        ifStop = true;
+        break;
+      default:
+        break;
+    }
+    if (ifStop) {
+      event.stopPropagation();
+    }
+  }
```

Comment:
lets try to include case's for the angular cdk keycodes as well when handling keydowns

`import {
  DOWN_ARROW,
  END,
  ENTER,
  HOME,
  LEFT_ARROW,
  PAGE_DOWN,
  PAGE_UP,
  RIGHT_ARROW,
  SPACE,
  UP_ARROW,
} from '@angular/cdk/keycodes';`

Reply:
updated